src/Executor/Promise/Adapter/SyncPromise.php

```diff
@@ -143,34 +179,54 @@ private function enqueueWaitingPromises(): void
             throw new InvariantViolation('Cannot enqueue derived promises when parent is still pending');
         }
 
-        foreach ($this->waiting as $descriptor) {
-            self::getQueue()->enqueue(function () use ($descriptor): void {
-                [$promise, $onFulfilled, $onRejected] = $descriptor;
-
-                if ($this->state === self::FULFILLED) {
-                    try {
-                        $promise->resolve($onFulfilled === null ? $this->result : $onFulfilled($this->result));
-                    } catch (\Throwable $e) {
-                        $promise->reject($e);
-                    }
-                } elseif ($this->state === self::REJECTED) {
-                    try {
-                        if ($onRejected === null) {
-                            $promise->reject($this->result);
-                        } else {
-                            $promise->resolve($onRejected($this->result));
-                        }
-                    } catch (\Throwable $e) {
-                        $promise->reject($e);
-                    }
-                }
-            });
+        // Capture state and result as values rather than $this reference to reduce memory per queued item
+        $state = $this->state;
+        $result = $this->result;
+        $queue = self::getQueue();
+
+        foreach ($this->waiting as $promise) {
+            $promise->waitingState = $state;
+            $promise->waitingResult = $result;
+            $queue->enqueue($promise);
         }
 
         $this->waiting = [];
     }
 
-    /** @return \SplQueue<callable(): void> */
+    /**
+     * Process a waiting promise task from the queue.
+     *
+     * @throws \Exception
+     */
+    private static function processWaitingTask(self $task): void
+    {
+        $onFulfilled = $task->waitingOnFulfilled;
+        $onRejected = $task->waitingOnRejected;
+        $state = $task->waitingState;
+        $result = $task->waitingResult;
+
+        // Clear references to allow GC
+        $task->waitingOnFulfilled = null;
+        $task->waitingOnRejected = null;
+        $task->waitingState = null;
+        $task->waitingResult = null;
+
+        try {
+            if ($state === self::FULFILLED) {
+                $task->resolve($onFulfilled === null ? $result : $onFulfilled($result));
```

Comment:
I think this would be easier to understand as:


```suggestion
                if ($onFulfilled) {
                    $result = $onFulfilled($result);
                }
                $task->resolve($result);
```

More literal code is fine here, since PHP will optimize the loop when it compiles.